[FLINK-1550/FLINK-4057] Add JobManager Metrics

zentol opened the pull request:
This PR relies on #2145.

This PR exposes several Metrics on the JobManager.

The first change made was the separation of the "metrics.jmx.port" key into 2 version, one 1 Task- and JobManagers respectively. ("(job/task)manager.metrics.jmx.port") This was necessary since otherwise you could never access TM metrics in a local setup.

Next up the metric infrastructure was introduced in the JM.

This includes 
* defining a new Scope format key (metrics.scope.jm) and default Scope format
* add a JobManagerMetricGroup
* refactoring the old TaskManager(Job)Group so that the JobMetricGroup can be used by both JM and TM
* creating/teardown MetricRegistry/JobManagerMetricGroup in the JobManager

Now metrics could be added.

In the JobManager scope we have:
* JMV metrics, which are a copy&paste from the TM
* availableSlots, totalSlots
* numRegisteredTaskManagers
* numRunningJobs

Adding these was straight-forward.

In the Job scope we have
* lastCheckpointSize
* lastCheckpointDuration

For this we have to pass a metricGroup through the ExecutionGraph into the CheckpointCoordinator.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/metrics/groups/JobManagerJobMetricGroup.java:
```diff
@@ -0,0 +1,67 @@
+ * Special {@link org.apache.flink.metrics.MetricGroup} representing everything belonging to
+ * a specific job, running on the JobManager.
+ */
+public class JobManagerJobMetricGroup extends JobMetricGroup {
```
Isn't this internal?

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/metrics/groups/JobManagerMetricGroup.java:
```diff
@@ -0,0 +1,100 @@
+		synchronized (this) {
+			currentJobGroup = jobs.get(jobId);
+
+			if (currentJobGroup == null || currentJobGroup.isClosed()) {
```
Can it happen that a `JobManagerJobMetricGroup` is closed and still kept in the `jobs` map?

tillrohrmann replied:
If not, then we could replace the HashMap with a `ConcurrentHashMap` and get rid of the `JobManagerJobMetricGroup`-wide locking.

zentol replied:
technically yes, it can be closed but still in the map. That is if someone messes up and adds that call to the JM.

Under usual circumstances no, the jobs map will never contain closed groups.

zentol replied:
Also, i would prefer it to stay the way it is. The MetricGroup's addX methods are already a wild west, I wouldn't want to add yet another way of handling stuff.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/metrics/groups/JobManagerMetricGroup.java:
```diff
@@ -0,0 +1,100 @@
+		// get or create a jobs metric group
+		JobManagerJobMetricGroup currentJobGroup;
+		synchronized (this) {
+			currentJobGroup = jobs.get(jobId);
```
Shouldn't we check whether the `JobManagerMetricGroup` is closed like we do it for the `TaskManagerMetricGroup`?

zentol replied:
yes

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/metrics/groups/TaskManagerJobMetricGroup.java:
```diff
@@ -0,0 +1,122 @@
+		}
+
+		// IMPORTANT: removing from the parent must happen while holding the this group's lock,
+		//      because it would violate the "first parent then subgroup" lock acquisition order
```
The comment seems to be not in sync with the implementation. Did you mean "must not happen"?

zentol replied:
i don't know, that's work from @StephanEwen

tillrohrmann replied:
I would assume that it is a typo. The `TaskManagerMetricGroup.removeJobMetricsGroup` does not access the `TaskManagerJobMetricGroup` where it could try to acquire the lock for the task manager job metric group.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/metrics/groups/scope/ScopeFormat.java:
```diff
@@ -66,23 +68,44 @@
+	/** The default scope format for the job component: {@code "<job_name>"} */
+	public static final String DEFAULT_SCOPE_JOBMANAGER_JOB_COMPONENT = SCOPE_JOB_NAME;
+
+	/** The default scope format for all job metrics: {@code "<host>.taskmanager.<tm_id>.<job_name>"} */
```
Shouldn't it be `<host>.jobmanager.<job_name>`?

zentol replied:
yes

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/metrics/groups/scope/ScopeFormat.java:
```diff
@@ -240,7 +305,7 @@ public OperatorScopeFormat(String format, TaskScopeFormat parentFormat) {
 			final String[] values = {
 					parent.parent().parent().hostname(),
-					parent.parent().parent().taskManagerId(),
+					((TaskManagerMetricGroup) parent.parent().parent()).taskManagerId(),
```
Redundant cast.

zentol replied:
relict of previous iterations, will remove it.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -78,6 +82,7 @@ import org.apache.flink.util.{InstantiationUtil, NetUtils}
 import org.jboss.netty.channel.ChannelException
 
 import scala.annotation.tailrec
+import scala.collection.JavaConversions._
```
The usage of implicit conversion via `JavaConversions` is discouraged. Better use `JavaConverters'` explicit conversion methods which you call via `.asScala`

zentol replied:
Took that line from the TaskManager.scala

tillrohrmann replied:
Then it should be refactored there as well.

zentol replied:
I'll change it in the JM. Since other lines in the TM may rely on that import this should imo be handled in another PR.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -149,6 +154,9 @@ class JobManager(
   var leaderSessionID: Option[UUID] = None
 
+  private var metricsRegistry : FlinkMetricRegistry = _
+  private var jobManagerMetricGroup : JobManagerMetricGroup = _
```
`null` values are discouraged in Scala. Better use `Option[FlinkMetricRegistry]` which you initialize with `None`.

tillrohrmann replied:
We could maybe create the `JobManagerMetricGroup` directly in the constructor of the `JobManager` if we pass in the `FlinkMetricRegistry` as a constructor argument.

zentol replied:
i know about Options but @StephanEwen changed it.

zentol replied:
Whoops, this issue was my bad actually.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -149,6 +154,9 @@ class JobManager(
 
   var leaderSessionID: Option[UUID] = None
 
+  private var metricsRegistry : FlinkMetricRegistry = _
```
So far, we've created the JobManager's components outside of the JobManager and passed them to it as a constructor argument. This had the advantage that if one of the components failed, it would not cause the `JobManager` to fail and the actor to die. But instead we could fail early telling exactly where the problem occurred.

Maybe we could do the same for the `FlinkMetricRegistry`.

zentol replied:
In theory this shouldn't be necessary as all metric-related code should never be able to crash anything.

tillrohrmann replied:
But it would be consistent with how we create components for the JobManager.

zentol replied:
that is true. also, it's just theory :)

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -270,6 +284,17 @@ class JobManager(
 
+    // failsafe shutdown of the metrics registry
+    try {
+      val reg = metricsRegistry
```
Why do you assign `metricsRegistry` to `reg`?

zentol replied:
Dunno, code from @StephanEwen

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1737,6 +1766,149 @@ class JobManager(
+    val direct = metrics.addGroup("Direct")
+    direct.gauge("Count", new Gauge[Long] {
+      override def getValue: Long = con
+        .getAttribute(directObjectName, "Count").asInstanceOf[Long]
```
Are we sure that `.getAttribute` always returns a non `null` value?

tillrohrmann replied:
But it throws exceptions when the attribute cannot be found. We should handle them.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1737,6 +1766,149 @@ class JobManager(
+        .getAttribute(directObjectName, "TotalCapacity").asInstanceOf[Long]
+    })
+
+    val mappedObjectName = new ObjectName("java.nio:type=BufferPool,name=direct")
```
Is it correct that `mappedObjectName.equals(directObjectName)`?

zentol replied:
no

tillrohrmann commented:
Really good work @zentol. I had some minor comments concerning the instantiation of the `MetricRegistry` for the `JobManager` and some clarifications. I think it would be good if we could add some test cases which check, for example, that we report the correct checkpoint size and duration.

Maybe you could also rebase this PR on the referred PR #2145. Then it would be easier to test the job manager metrics on a local machine. I would like to do it before merging. But once this is done, then I think we're good to merge your PR :-)

zentol commented:
@tillrohrmann I've addressed your comments and rebased the PR. You should now be able to try it out locally.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -149,6 +154,16 @@ class JobManager(
+      Option(new JobManagerMetricGroup(
+        registry, NetUtils.ipAddressToUrlString(InetAddress.getByName(host))))
+    case None =>
+      log.error("Could not instantiate JobManager metrics.")
```
Is this an error or more of a warning? I mean the system would still run, wouldn't it?

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -204,6 +219,12 @@ class JobManager(
+      case Some(group) =>
+        instantiateMetrics(group)
+      case None =>
+        log.error("Could not instantiate JobManager metrics.")
```
error or warning?

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -2325,55 +2326,65 @@ object TaskManager {
           .getAttribute(directObjectName, "TotalCapacity").asInstanceOf[Long]
       })
+    } catch {
+      case e: Exception =>
```
Which exception are you catching here? Shouldn't they surround the `con.getAttribute` call in the gauges?

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1727,6 +1764,160 @@ class JobManager(
+    try {
+      direct.gauge("Count", new Gauge[Long] {
+        override def getValue: Long = con
+          .getAttribute(directObjectName, "Count").asInstanceOf[Long]
```
I think the try-catch block should surround the `con.getAttribute` call instead of the calls which add the gauge to the metrics groups.

zentol replied:
ehh....yeah.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1082,6 +1110,13 @@ class JobManager(
 
+        val jobMetrics = jobManagerMetricGroup match {
+          case Some(group) =>
+            group.addJob(jobGraph.getJobID, jobGraph.getName)
```
What happens if `group.addJob` returns a `null` value? Then the instantiation of the `ExecutionGraph` would fail with a `NullPointerException`. We should check that and return an `UnregisteredMetricsGroup` instead.

zentol replied:
Do you think returning null is a good pattern in the first place? I'd rather return a jobGroup that is neither stored in the jobManagerMetricGroup  nor registers any metrics.

tillrohrmann replied:
That works as well and is probably safer.

I actually just run into this problem. Have you actually tried your changes out with submitting a job?

zentol replied:
most of them, and always when making major changes.

But not after adding the isClosed() check to the jobManagerMetricGroup apparently.

tillrohrmann commented:
Good fixes @zentol. I had some minor comments which we should address before merging this PR.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/metrics/MetricRegistry.java:
```diff
@@ -30,6 +30,7 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import javax.management.JMX;
```
Unused import

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1727,6 +1764,160 @@ class JobManager(
+
+    val mapped = metrics.addGroup("Mapped")
+    try {
+      mapped.gauge("Count", new Gauge[Long] {
```
What does this metric tell me?

zentol replied:
the number of buffers in the pool. the metric names were taken from https://docs.oracle.com/javase/7/docs/api/java/lang/management/BufferPoolMXBean.html

tillrohrmann replied:
Thanks for the info :-)

tillrohrmann commented:
I just tried the code out and run into the problem of a job metric group being null:

```
java.lang.NullPointerException
	at org.apache.flink.util.Preconditions.checkNotNull(Preconditions.java:58)
	at org.apache.flink.runtime.executiongraph.ExecutionGraph.<init>(ExecutionGraph.java:316)
	at org.apache.flink.runtime.jobmanager.JobManager.org$apache$flink$runtime$jobmanager$JobManager$$submitJob(JobManager.scala:1126)
	at org.apache.flink.runtime.jobmanager.JobManager$$anonfun$handleMessage$1.applyOrElse(JobManager.scala:501)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)
	at org.apache.flink.runtime.LeaderSessionMessageFilter$$anonfun$receive$1.applyOrElse(LeaderSessionMessageFilter.scala:36)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply$mcVL$sp(AbstractPartialFunction.scala:33)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:33)
	at scala.runtime.AbstractPartialFunction$mcVL$sp.apply(AbstractPartialFunction.scala:25)
	at org.apache.flink.runtime.LogMessages$$anon$1.apply(LogMessages.scala:33)
	at org.apache.flink.runtime.LogMessages$$anon$1.apply(LogMessages.scala:28)
	at scala.PartialFunction$class.applyOrElse(PartialFunction.scala:118)
	at org.apache.flink.runtime.LogMessages$$anon$1.applyOrElse(LogMessages.scala:28)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
	at org.apache.flink.runtime.jobmanager.JobManager.aroundReceive(JobManager.scala:118)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
	at akka.dispatch.Mailbox.run(Mailbox.scala:221)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

zentol commented:
fixed the issue with the group being null.

tillrohrmann commented:
Could we maybe add a test case where we try to obtain the checkpoint duration or size or another metric from the `JobManager`? This would guard against future changes to the job metric group, for example.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -663,6 +677,9 @@ public boolean receiveAcknowledgeMessage(AcknowledgeCheckpoint message) throws E
 								completed.getDuration() + " ms)");
 
+						lastCheckpointSize = completed.getStateSize();
+						lastCheckpointDuration = completed.getDuration();
```
Would it make sense to use the `CheckpointStatsTracker` to obtain the values for the last checkpoint? Given that the JobManager is not so high performance critical with respect to the calculation of the metrics.

tillrohrmann commented:
I've tested the PR locally and the JobManager metrics were shown properly. Good job @zentol :-)

Thus, there are only a test case for the job manager metrics, the question whether to use the `CheckpointStatsTracker` or not for the checkpoint metrics and the try-catch blocks around the `con.getAttribute` calls left.

zentol commented:
I moved the checkpoint metrics into the Tracker (and reverted the changed to ExecutionGraph). Currently trying it out locally.

Regarding the exception catching in the metrics: I can't decide whether we should try to write all metrics in such a way that they can't throw exceptions, or write the reporters in such a way that they can deal with it. (usually by logging the exception). The first option is safer considering custom reporters, but the second will allows us to properly log them.

Regarding a test: While i agree that such a test would be nice i can only come up with icky ways to test it. You have to access the metric _while a job is running_ as they are removed afterwards. So you either have to submit a job that blocks until _something_ happens, or you add a reporter that feeds that information back to the test _somehow_.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/stats/SimpleCheckpointStatsTracker.java:
```diff
@@ -106,10 +108,12 @@
 	 * The latest completed checkpoint (highest ID) or <code>null</code>.
 	 */
 	private CompletedCheckpoint latestCompletedCheckpoint;
+	private CheckpointStats latestCompletedCheckpointStats;
```
Where is this variable set?

zentol replied:
this variable is unused.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1235,7 +1234,7 @@ class JobManager(
                 ConfigConstants.DEFAULT_JOB_MANAGER_WEB_CHECKPOINTS_HISTORY_SIZE)
 
-              new SimpleCheckpointStatsTracker(historySize, ackVertices)
+              new SimpleCheckpointStatsTracker(historySize, ackVertices, jobMetrics)
```
Where do we close the `jobMetrics` group?

tillrohrmann replied:
We only have to create a job metric group if we also create a `SimplCheckpointStatsTracker`.

zentol replied:
JobManager.scala, line 1664.

I would also create it regardless of the Tracker. The moment we use it in another component we would have to move it out of that condition anyways.

tillrohrmann replied:
But then we should do it when it becomes necessary and not now.

tillrohrmann commented:
I think it is good to integrate the metrics into the `CheckpointStatsTracker`.

Even if such a test seems to be icky, I think it would be a valuable addition. We already use this kind of pattern for a multitude of other tests and it has proven to work.

zentol commented:
Can you point me to a test that does this?

tillrohrmann commented:
The `JobManagerHARecoveryTest.testJobRecoveryWhenLosingLeadership` uses a `BlockingInvokable`, for example.

zentol commented:
@tillrohrmann I've added a test case

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerMetricTest.java:
```diff
@@ -0,0 +1,249 @@
+			myLeaderRetrievalService.notifyListener(gateway.path(), leaderSessionID);
+
+			Await.ready(isLeader, deadline.timeLeft());
+			Await.ready(isConnectedToJobManager, deadline.timeLeft());
```
Is there a specific reason why you start the cluster components individually? I think you could also use the `TestingCluster` which starts a cluster of `TestingJobManager` and `TestingTaskManager`.

zentol replied:
I simply took the test case you pointed me to and adjusted it for my use-case.

tillrohrmann replied:
Hmm I though you were only looking for a blocking `Invokable` implementation. I guess it works how you did it, but it doesn't look that nice because of all the explicit steps.

zentol replied:
I agree, I'll look into the TestCluster stuff

zentol replied:
I've refactored the test case.

tillrohrmann replied:
Can it be that you haven't pushed you commit yet? Because I still see the old version.

zentol replied:
sorry, git was being stupid. It's pushed now.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerMetricTest.java:
```diff
@@ -0,0 +1,249 @@
+
+			MBeanServer mBeanServer = ManagementFactory.getPlatformMBeanServer();
+			ObjectName objectName1 = new ObjectName("org.apache.flink.metrics:key0=jobmanager,key1=TestingJob,name=lastCheckpointSize");
+			assertEquals(-1L, mBeanServer.getAttribute(objectName1, "Value"));
```
Do we receive `-1` because no checkpoint has been completed yet? Could we maybe complete a checkpoint which has size `0`?

zentol replied:
-1 is the value that gauges return when no checkpoint was completed yet / lastCompletedCheckpoint in StatsTracker = null.

Of course we could, but the purpose of this test was to make sure the instantiation of the jobMetricGroup and metrics is done correctly, which it verifies.

zentol commented:
That is a known and intended behavior. It is the users responsibility to make sure that no naming conflicts arise, either by providing unique job/task/operator names or modifying the scope formats.

Regarding your suggestion, even if we exclude the disparity between the name a user expects and the name we might provide, there is still the issue that it (sadly) doesn't actually solve anything.

Assume 2 TaskManagers, both exporting a metric `MyJob.metric` and `MyJob1.metric`.

You cannot guarantee that both versions of `MyJob.metric` actually belong to the same job.
You can't even make the guarantee that all values a TM exposes as `MyJob.metric` belong to a single job.

This problem also exists right; the only difference is that you only have 1 set of inconsistent data, and not 2.

tillrohrmann commented:
Hmm what about using the `JobID` to disambiguate jobs with the same name?

I cannot remember whether this peculiarity concerning the job naming is documented or not. If not, then it should be included in the metrics documentation PR.

zentol commented:
What is this discussion about; changing the default format or finding a general way to avoid naming collisions?

zentol commented:
For clarification: including the JobID is exactly what the user is supposed to do when he runs multiple jobs with the same name.

The current scope format is designed to provide the most readable name. As such it includes as few ID's and as many names as possible. Could we change it? Sure. But that's a separate issue and should in no way be part of this PR.

tillrohrmann commented:
Alright. I just reported my observation while trying this PR out. 

You're right that it would blow up the scope of this PR. I guess the easiest solution would be to properly document the behaviour when a user submits jobs with the same name and to tell him that he should rather use the job id in this case.

tillrohrmann commented:
Perfect :-) Thanks for your work @zentol. I think the PR is now good to be merged.

zentol commented:
Merging.
